subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/FileCollectionCodec.kt

```diff
@@ -137,6 +142,15 @@ class CollectingVisitor : FileCollectionStructureVisitor {
                 elements.add(FilteredFileCollectionSpec(fileCollection.collection, fileCollection.filterSpec))
                 false
             }
+            is ProviderBackedFileCollection -> {
+                val provider = fileCollection.provider
+                if (provider !is TaskProvider<*>) {
```

Comment:
What is the check for? Perhaps add a comment to explain why it's there (so we can later refactor things to get rid of it).

Reply:
Good point. The check is there to guard against file collections created directly from task providers such as `layout.files(compileJava)` then referenced from a different task.

Reply:
Done in https://github.com/gradle/gradle/pull/14348.